window-purpose-core.el

```diff
@@ -79,12 +79,13 @@ dummy buffer with the purpose 'edit."
 mode and MODE-CONF.
 MODE-CONF is a hash table mapping modes to purposes."
   (when (get-buffer buffer-or-name)     ; check if buffer exists
-    (let* ((major-mode (purpose--buffer-major-mode buffer-or-name))
-           (derived-modes (purpose--iter-hash #'(lambda (mode _purpose) mode)
-                                              mode-conf))
-           (derived-mode (apply #'derived-mode-p derived-modes)))
-      (when derived-mode
-        (gethash derived-mode mode-conf)))))
+    (cl-block nil
+      (maphash
+       (let ((buffer-mode (purpose--buffer-major-mode buffer-or-name)))
+         #'(lambda (mode purpose)
+             (when (provided-mode-derived-p buffer-mode mode)
+               (cl-return purpose))))
+       mode-conf))))
```

Comment:
The problem with this is that a buffer's major mode can derive from multiple modes. For example, if mode `a-mode` derives from `b-mode` and `a-mode` is configured to have purpose `A` and `b-mode` is configured to have purpose `B`, then your code can return either `A` or `B`, depending on which mode is listed first in `mode-conf`. The correct behavior in this case is to return `A`, and the old code guarantees that (because `derived-mode-p` returns the closest parent).

This change needs to be reverted.

Reply:
Also, `provided-mode-derived-p` was introduced in Emacs 26.1, so we can't use it yet (I still try to maintain compatibility with Emacs 24.4).

Reply:
That's a very good point. I had thought of this too, but mistakenly thought it had the same behaviour. 
To keep it simplest I'll just revert this whole chunk since 98% of the fixes to the underlying consing issue are the changes in `purpose-x-replace-buffer-in-windows-1`.